### PR TITLE
Auto-detect SAM CLI if not set manually

### PR DESCRIPTION
**Pull request by abrooksv**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description
Add a SAM CLI detector that looks in common OS-specific locations if the customer did not configure it manually

## Related Issue(s)
#253

## Screenshots (if appropriate)
screen shot 2018-11-06 at 1 53 06 pm

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: core/tst/software/aws/toolkits/core/rules/EnvironmentVariableHelper.kt [hunk 1]
```diff
@@ -22,28 +22,33 @@ class EnvironmentVariableHelper : ExternalResource() {
         keys.forEach { modifiableMap.remove(it) }
     }
 
-    operator fun set(key: String, value: String) {
+    operator fun set(key: String, value: String?) {
```
Comment (kiiadi) [inline core/tst/software/aws/toolkits/core/rules/EnvironmentVariableHelper.kt]: Why not have an explicit remove?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/settings/SamExecutableDetector.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.settings
+
+import com.intellij.openapi.util.SystemInfo
+import com.intellij.openapi.util.text.StringUtil
+import java.io.File
+
+open class SamExecutableDetector {
+    private companion object {
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/settings/SamExecutableDetector.kt]: Why is this at the top? You disgust me.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/settings/SamExecutableDetector.kt [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    fun detect(): String? = if (SystemInfo.isWindows) {
+        detectForWindows()
+    } else {
+        detectForUnix()
+    }
+
+    private fun detectForWindows(): String? {
+        for (path in WINDOWS_PATHS) {
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/settings/SamExecutableDetector.kt]: why the classic for loops and not a `foreach` ?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/settings/SamExecutableDetector.kt (hunk 2 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/settings/SamExecutableDetector.kt]: I disgust myself 🤢

File: core/tst/software/aws/toolkits/core/rules/EnvironmentVariableHelper.kt (hunk 1 above)
Comment (abrooksv) [reply core/tst/software/aws/toolkits/core/rules/EnvironmentVariableHelper.kt]: Sure

File: jetbrains-core/src/software/aws/toolkits/jetbrains/settings/SamExecutableDetector.kt (hunk 3 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/settings/SamExecutableDetector.kt]: Replaced

File: jetbrains-core/src/software/aws/toolkits/jetbrains/settings/SamExecutableDetector.kt (hunk 2 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/settings/SamExecutableDetector.kt]: 😆

**Pull request merged**